ClassNotFoundException: org.springframework.boot.webclient.WebClientCustomizer

See stacktrace below. This is on Spring Boot 4.

https://github.com/belgif/rest-problem-java/blob/8b19e2f6548fafa8c6f418d19e77b8652588c667/belgif-rest-problem-spring-boot-4-client-starter/src/main/java/io/github/belgif/rest/problem/spring/client/ClientProblemAutoConfiguration.java#L43-L47

```
java.lang.IllegalStateException: Failed to introspect Class [io.github.belgif.rest.problem.spring.client.ClientProblemAutoConfiguration] from ClassLoader [jdk.internal.loader.ClassLoaders$AppClassLoader@76ed5528]
	at org.springframework.util.ReflectionUtils.getDeclaredMethods(ReflectionUtils.java:483)
	at org.springframework.util.ReflectionUtils.doWithMethods(ReflectionUtils.java:360)
	at org.springframework.util.ReflectionUtils.getUniqueDeclaredMethods(ReflectionUtils.java:417)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.lambda$getTypeForFactoryMethod$0(AbstractAutowireCapableBeanFactory.java:750)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1740)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.getTypeForFactoryMethod(AbstractAutowireCapableBeanFactory.java:749)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.determineTargetType(AbstractAutowireCapableBeanFactory.java:683)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.predictBeanType(AbstractAutowireCapableBeanFactory.java:655)
	at org.springframework.beans.factory.support.AbstractBeanFactory.isFactoryBean(AbstractBeanFactory.java:1699)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doGetBeanNamesForType(DefaultListableBeanFactory.java:632)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBeanNamesForType(DefaultListableBeanFactory.java:604)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBeanNamesForType(DefaultListableBeanFactory.java:589)
	at org.springframework.boot.autoconfigure.condition.OnBeanCondition.collectBeanDefinitionsForType(OnBeanCondition.java:334)
	at org.springframework.boot.autoconfigure.condition.OnBeanCondition.getBeanDefinitionsForType(OnBeanCondition.java:323)
	at org.springframework.boot.autoconfigure.condition.OnBeanCondition.getMatchingBeans(OnBeanCondition.java:221)
	at org.springframework.boot.autoconfigure.condition.OnBeanCondition.evaluateConditionalOnMissingBean(OnBeanCondition.java:204)
	at org.springframework.boot.autoconfigure.condition.OnBeanCondition.getMatchOutcome(OnBeanCondition.java:149)
	at org.springframework.boot.autoconfigure.condition.SpringBootCondition.matches(SpringBootCondition.java:47)
	... 94 common frames omitted
Caused by: java.lang.NoClassDefFoundError: org/springframework/boot/webclient/WebClientCustomizer
	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1027)
	at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:150)
	at java.base/jdk.internal.loader.BuiltinClassLoader.defineClass(BuiltinClassLoader.java:862)
	at java.base/jdk.internal.loader.BuiltinClassLoader.findClassOnClassPathOrNull(BuiltinClassLoader.java:760)
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClassOrNull(BuiltinClassLoader.java:681)
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:639)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526)
	at java.base/java.lang.Class.getDeclaredMethods0(Native Method)
	at java.base/java.lang.Class.privateGetDeclaredMethods(Class.java:3578)
	at java.base/java.lang.Class.getDeclaredMethods(Class.java:2676)
	at org.springframework.util.ReflectionUtils.getDeclaredMethods(ReflectionUtils.java:465)
	... 111 common frames omitted
Caused by: java.lang.ClassNotFoundException: org.springframework.boot.webclient.WebClientCustomizer
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526)
	... 124 common frames omitted
```

--- a/belgif-rest-problem-spring-boot-3-client-starter/src/main/java/io/github/belgif/rest/problem/spring/client/ClientProblemAutoConfiguration.java
+++ b/belgif-rest-problem-spring-boot-3-client-starter/src/main/java/io/github/belgif/rest/problem/spring/client/ClientProblemAutoConfiguration.java
@@ -29,22 +29,29 @@ public class ClientProblemAutoConfiguration {
     }
 
     @ConditionalOnClass({ RestClient.class, RestClientCustomizer.class })
-    @Bean
-    public ProblemRestClientCustomizer
-            problemRestClientCustomizer(ProblemResponseErrorHandler problemResponseErrorHandler) {
-        return new ProblemRestClientCustomizer(problemResponseErrorHandler);
+    public static class RestClientProblemConfiguration {
+        @Bean
+        public ProblemRestClientCustomizer
+                problemRestClientCustomizer(ProblemResponseErrorHandler problemResponseErrorHandler) {
+            return new ProblemRestClientCustomizer(problemResponseErrorHandler);
+        }
     }
 
     @ConditionalOnClass({ RestTemplate.class, RestTemplateCustomizer.class })
-    @Bean
-    public ProblemRestTemplateCustomizer
-            problemRestTemplateCustomizer(ProblemResponseErrorHandler problemResponseErrorHandler) {
-        return new ProblemRestTemplateCustomizer(problemResponseErrorHandler);
+    public static class RestTemplateProblemConfiguration {
+        @Bean
+        public ProblemRestTemplateCustomizer
+                problemRestTemplateCustomizer(ProblemResponseErrorHandler problemResponseErrorHandler) {
+            return new ProblemRestTemplateCustomizer(problemResponseErrorHandler);
+        }
     }
 
     @ConditionalOnClass({ WebClient.class, WebClientCustomizer.class })
-    @Bean
-    public ProblemWebClientCustomizer problemWebClientCustomizer() {
-        return new ProblemWebClientCustomizer();
+    public static class WebClientProblemConfiguration {
+        @Bean
+        public ProblemWebClientCustomizer problemWebClientCustomizer() {
+            return new ProblemWebClientCustomizer();
+        }
     }
+
 }

--- a/belgif-rest-problem-spring-boot-3-server-starter/src/main/java/io/github/belgif/rest/problem/spring/server/ServerProblemAutoConfiguration.java
+++ b/belgif-rest-problem-spring-boot-3-server-starter/src/main/java/io/github/belgif/rest/problem/spring/server/ServerProblemAutoConfiguration.java
@@ -22,18 +22,6 @@ import com.fasterxml.jackson.databind.ObjectMapper;
 @Import({ io.github.belgif.rest.problem.spring.ProblemJackson2Configuration.class, JacksonAutoConfiguration.class })
 public class ServerProblemAutoConfiguration {
 
-    private ObjectMapper objectMapper;
-
-    public ServerProblemAutoConfiguration(ObjectMapper objectMapper) {
-        this.objectMapper = objectMapper;
-    }
-
-    @ConditionalOnClass(ConstraintViolationException.class)
-    @Bean
-    public BeanValidationExceptionsHandler beanValidationExceptionsHandler() {
-        return new BeanValidationExceptionsHandler();
-    }
-
     @Bean
     public ProblemExceptionHandler problemExceptionHandler() {
         return new ProblemExceptionHandler();
@@ -44,15 +32,34 @@ public class ServerProblemAutoConfiguration {
         return new RoutingExceptionsJackson2Handler();
     }
 
+    @ConditionalOnClass(ConstraintViolationException.class)
+    public static class BeanValidationProblemConfiguration {
+        @Bean
+        public BeanValidationExceptionsHandler beanValidationExceptionsHandler() {
+            return new BeanValidationExceptionsHandler();
+        }
+    }
+
     @ConditionalOnClass({ Configuration.class, ValidationConfigurationCustomizer.class })
-    @Bean
-    public ProblemValidationConfigurationCustomizer problemValidationConfigurationCustomizer() {
-        return new ProblemValidationConfigurationCustomizer();
+    public static class BeanValidationConfigurationProblemConfiguration {
+        @Bean
+        public ProblemValidationConfigurationCustomizer problemValidationConfigurationCustomizer() {
+            return new ProblemValidationConfigurationCustomizer();
+        }
     }
 
     @ConditionalOnClass(InvalidRequestException.class)
-    @Bean
-    public InvalidRequestExceptionJackson2Handler invalidRequestExceptionHandler() {
-        return new InvalidRequestExceptionJackson2Handler(objectMapper);
+    public static class SwaggerValidatorProblemConfiguration {
+        private final ObjectMapper objectMapper;
+
+        public SwaggerValidatorProblemConfiguration(ObjectMapper objectMapper) {
+            this.objectMapper = objectMapper;
+        }
+
+        @Bean
+        public InvalidRequestExceptionJackson2Handler invalidRequestExceptionHandler() {
+            return new InvalidRequestExceptionJackson2Handler(objectMapper);
+        }
     }
+
 }

--- a/belgif-rest-problem-spring-boot-4-client-starter/src/main/java/io/github/belgif/rest/problem/spring/client/ClientProblemAutoConfiguration.java
+++ b/belgif-rest-problem-spring-boot-4-client-starter/src/main/java/io/github/belgif/rest/problem/spring/client/ClientProblemAutoConfiguration.java
@@ -27,22 +27,29 @@ public class ClientProblemAutoConfiguration {
     }
 
     @ConditionalOnClass({ RestClient.class, RestClientCustomizer.class })
-    @Bean
-    public ProblemRestClientCustomizer
-            problemRestClientCustomizer(ProblemResponseErrorHandler problemResponseErrorHandler) {
-        return new ProblemRestClientCustomizer(problemResponseErrorHandler);
+    public static class RestClientProblemConfiguration {
+        @Bean
+        public ProblemRestClientCustomizer
+                problemRestClientCustomizer(ProblemResponseErrorHandler problemResponseErrorHandler) {
+            return new ProblemRestClientCustomizer(problemResponseErrorHandler);
+        }
     }
 
     @ConditionalOnClass({ RestTemplate.class, RestTemplateCustomizer.class })
-    @Bean
-    public ProblemRestTemplateCustomizer
-            problemRestTemplateCustomizer(ProblemResponseErrorHandler problemResponseErrorHandler) {
-        return new ProblemRestTemplateCustomizer(problemResponseErrorHandler);
+    public static class RestTemplateProblemConfiguration {
+        @Bean
+        public ProblemRestTemplateCustomizer
+                problemRestTemplateCustomizer(ProblemResponseErrorHandler problemResponseErrorHandler) {
+            return new ProblemRestTemplateCustomizer(problemResponseErrorHandler);
+        }
     }
 
     @ConditionalOnClass({ WebClient.class, WebClientCustomizer.class })
-    @Bean
-    public ProblemWebClientCustomizer problemWebClientCustomizer() {
-        return new ProblemWebClientCustomizer();
+    public static class WebClientProblemConfiguration {
+        @Bean
+        public ProblemWebClientCustomizer problemWebClientCustomizer() {
+            return new ProblemWebClientCustomizer();
+        }
     }
+
 }

--- a/belgif-rest-problem-spring-boot-4-server-starter/src/main/java/io/github/belgif/rest/problem/spring/server/ServerProblemAutoConfiguration.java
+++ b/belgif-rest-problem-spring-boot-4-server-starter/src/main/java/io/github/belgif/rest/problem/spring/server/ServerProblemAutoConfiguration.java
@@ -23,18 +23,6 @@ import tools.jackson.databind.ObjectMapper;
 @Import({ io.github.belgif.rest.problem.spring.ProblemJackson3Configuration.class, JacksonAutoConfiguration.class })
 public class ServerProblemAutoConfiguration {
 
-    private ObjectMapper objectMapper;
-
-    public ServerProblemAutoConfiguration(ObjectMapper objectMapper) {
-        this.objectMapper = objectMapper;
-    }
-
-    @ConditionalOnClass(ConstraintViolationException.class)
-    @Bean
-    public BeanValidationExceptionsHandler beanValidationExceptionsHandler() {
-        return new BeanValidationExceptionsHandler();
-    }
-
     @Bean
     public ProblemExceptionHandler problemExceptionHandler() {
         return new ProblemExceptionHandler();
@@ -45,15 +33,34 @@ public class ServerProblemAutoConfiguration {
         return new RoutingExceptionsJackson3Handler();
     }
 
+    @ConditionalOnClass(ConstraintViolationException.class)
+    public static class BeanValidationProblemConfiguration {
+        @Bean
+        public BeanValidationExceptionsHandler beanValidationExceptionsHandler() {
+            return new BeanValidationExceptionsHandler();
+        }
+    }
+
     @ConditionalOnClass({ Configuration.class, ValidationConfigurationCustomizer.class })
-    @Bean
-    public ProblemValidationConfigurationCustomizer problemValidationConfigurationCustomizer() {
-        return new ProblemValidationConfigurationCustomizer();
+    public static class BeanValidationConfigurationProblemConfiguration {
+        @Bean
+        public ProblemValidationConfigurationCustomizer problemValidationConfigurationCustomizer() {
+            return new ProblemValidationConfigurationCustomizer();
+        }
     }
 
     @ConditionalOnClass(InvalidRequestException.class)
-    @Bean
-    public InvalidRequestExceptionJackson3Handler invalidRequestExceptionHandler() {
-        return new InvalidRequestExceptionJackson3Handler(objectMapper);
+    public static class SwaggerRequestValidationConfiguration {
+        private final ObjectMapper objectMapper;
+
+        public SwaggerRequestValidationConfiguration(ObjectMapper objectMapper) {
+            this.objectMapper = objectMapper;
+        }
+
+        @Bean
+        public InvalidRequestExceptionJackson3Handler invalidRequestExceptionHandler() {
+            return new InvalidRequestExceptionJackson3Handler(objectMapper);
+        }
     }
+
 }

--- a/src/main/asciidoc/release-notes.adoc
+++ b/src/main/asciidoc/release-notes.adoc
@@ -12,6 +12,12 @@
 
 // tag::recent-versions[]
 
+== Version 0.22
+
+*belgif-rest-problem-spring-boot-**:
+
+* Fix `@ConditionalOnClass`  loading of optional components
+
 == Version 0.21
 
 *belgif-rest-problem:*

